File: Leetcode/Medium/199_binary_tree_right_side_view.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type node struct {
	*TreeNode
	level int
}

func RightSideView(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	result := make([]int, 0)
	queue := []node{{root, 0}}
	level := 0

	for len(queue) > 0 {
		head := queue[0]
		queue = queue[1:]
		if head.level == level {
			result = append(result, head.Val)
			level++
		}
		if head.Right != nil {
			queue = append(queue, node{head.Right, head.level + 1})
		}
		if head.Left != nil {
			queue = append(queue, node{head.Left, head.level + 1})
		}
	}

	return result
}

func main() {
	root := &TreeNode{Val: 2, Left: &TreeNode{Val: 1}}
	result := RightSideView(root)
	fmt.Println(result)
}

File: Leetcode/Medium/199_binary_tree_right_side_view.py
# Definition for a binary tree node.
from typing import Optional, List, Tuple


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        result: list[int] = []
        nodes: list[TreeNode] = [root]

        while True:
            nodes, level_val = self.helper(nodes)
            if len(level_val):
                result.append(level_val.pop())

            if not nodes:
                break

        return result

    def helper(self, nodes: list[TreeNode]) -> Tuple[list[TreeNode], list[int]]:
        level_val: list[int] = []
        level_nodes: list[TreeNode] = []
        for node in nodes:

            if node.left is not None:
                level_nodes.append(node.left)

            if node.right is not None:
                level_nodes.append(node.right)

            level_val.append(node.val)

        return level_nodes, level_val


if __name__ == "__main__":
    root = TreeNode(
        val=1,
        left=TreeNode(val=2, right=TreeNode(val=4)),
        right=TreeNode(val=3, right=TreeNode(val=5)),
    )

    solution = Solution()
    result = solution.rightSideView(root)
    print(result)
